R/recycle.R

```diff
@@ -1,31 +1,164 @@
+#' process used by packages in r-lib and the tidyverse:
+#'
+#' - Vectors of size 1 will be recycled to the size of any other vector
+#'
+#' - Otherwise, all vectors must have the same size
+#'
+#' @section Examples:
+#'
+#' Vectors of size 1 are recycled to the size of any other vector:
+#'
+#' ```
+#' tibble(x = 1:3, y = 1L)
+#' #> # A tibble: 3 × 2
+#' #>       x     y
+#' #>   <int> <int>
+#' #> 1     1     1
+#' #> 2     2     1
+#' #> 3     3     1
+#' ```
+#'
+#' This includes vectors of size 0:
+#'
+#' ```
+#' tibble(x = integer(), y = 1L)
+#' #> # A tibble: 0 × 2
+#' #> # … with 2 variables: x <int>, y <int>
+#' ```
+#'
+#' If vectors aren't size 1, they must all be the same size. Otherwise, an error
+#' is thrown:
+#'
+#' ```
+#' tibble(x = 1:3, y = 4:7)
+#' #> Error:
+#' #> ! Tibble columns must have compatible sizes.
+#' #> • Size 3: Existing data.
+#' #> • Size 4: Column `y`.
+#' #> ℹ Only values of size one are recycled.
+#' ```
+#'
+#' @section vctrs backend:
+#'
+#' Packages in r-lib and the tidyverse generally use [vec_size_common()] and
+#' [vec_recycle_common()] as the backends for handling recycling rules.
+#'
+#' - `vec_size_common()` returns the common size of multiple vectors, after
+#'   applying the recycling rules
+#'
+#' - `vec_recycle_common()` goes one step further, and actually recycles the
+#'   vectors to their common size
 #'
-#' `vec_recycle(x, size)` recycles a single vector to given size.
-#' `vec_recycle_common(...)` recycles multiple vectors to their common size.
-#' All functions obey the vctrs recycling rules, described below, and will
-#' throw an error if recycling is not possible. See [vec_size()] for the
-#' precise definition of size.
+#' ```
+#' vec_size_common(1:3, "x")
+#' #> [1] 3
 #'
-#' @section Recycling rules:
-#' The common size of two vectors defines the recycling rules, and can be
-#' summarise with the following table:
+#' vec_recycle_common(1:3, "x")
+#' #> [[1]]
+#' #> [1] 1 2 3
+#' #>
+#' #> [[2]]
+#' #> [1] "x" "x" "x"
 #'
-#' \figure{sizes-recycling.png}
+#' vec_size_common(1:3, c("x", "y"))
+#' #> Error:
+#' #> ! Can't recycle `..1` (size 3) to match `..2` (size 2).
+#' ```
 #'
-#' (Note `NULL`s are handled specially; they are treated like empty
-#' arguments and hence don't affect the size)
+#' @section Differences with base R:
 #'
-#' This is a stricter set of rules than base R, which will usually
-#' return output of length `max(nx, ny)`, warning if the length of the longer
-#' vector is not an integer multiple of the length of the shorter.
+#' The recycling rules described here are stricter than the ones generally used
+#' by base R. With base R, the rules are usually:
+#'
+#' - If any vector is length 0, the output will be length 0
+#'
+#' - Otherwise, the output will be length `max(length_x, length_y)`, and a
+#'   warning will be thrown if the length of the longer vector is not an integer
+#'   multiple of the length of the shorter vector.
+#'
+#' ```
```

Comment:
I think I'd drop the examples here. I think it makes sense to explores the inconsistencies in the vignette, but I'd keep it brief here since this is user facing. But we could say "See `vignette("...")` for more details".

Reply:
I've dropped the examples and linked to the vignette, but I've kept this as a small separate section at the very bottom of the doc. I think including the base R rules between our rules and the Examples section is a little confusing, and it looks better to me to include them at the very end